Adding I/O interfaces in GSI for Analysis of Significant Wave Height (HOWV) and near-surface Wind Gust (GUST) for WRF-ARW based 3DRTMA

GangZhao-NOAA opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**

<!-- Please include relevant motivation and context. -->
***Motivation:***

- The analysis of significant wave height (HOWV) and 10-m wind gust (GUST) are listed as standard product in the operational (2D)RTMA/URMA product suites. To keep providing these two products in next generation RTMA/URMA (ie, 3DRTMA) system, the capability to directly analyze wave height and wind gust must be added into GSI 3DVar analysis. 

***Method:***

- The kernel part of the analysis code (such as stphowv.f90, setuphowv.f90, inthowv.f90, gsi_howvOper.f90 and m_howvNode.f90 for HOWV, and similar code for GUST) for wave height and wind gust had already been introduced into GSI through the implementation of 2DRTMA. So most of the code work for analysis of HOWV and GUST in 3DVar, are mainly focused on the I/O interfaces for the firstguess provided by various models (i.e.  reading background of HOWV/GUST from firstguess file and writing the updated HOWV/GUST back to the final analysis file).
- The I/O interfaces for FV3-LAM model based firstguess had been done in previous work (see issue #601 and issue #726). Since RRFS-3DRTMAv1 has been switched to be based on HRRR model, so in this work, some efforts are made to add the I/O interfaces in GSI for WRF-ARW model based firstguess.
     
<!-- Please include a summary of the change and which issue is fixed. -->
- here is a [Google document](https://docs.google.com/document/d/1gprGzUv7sgXagSEXGHh4zYMCTAwjIbUdmFwsoUnwbkA/edit?usp=sharing) drafted as a brief note on the code changes.

<!-- List any dependencies that are required for this change. -->

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->
***Related Issues:***
- This Pull Request is to resolve issue #834 

Resolves #834 


**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published
- [x] Regression tests on WCOSS2 (Cactus) and Hera (Rocky-8) : my updated GSI commit [#47b3115](https://github.com/GangZhao-NOAA/GSI/commit/47b311534b8b1b0519edfbf3038e49c47aa5ccbe)) vs control/original GSI code (commit [#92165a4](https://github.com/NOAA-EMC/GSI/commit/92165a4242d41d4c29ec62a52f270e17c819fafd))
Here is the reports of the regression tests on WCOSS2 (Dogwood):
~~~
[gang.zhao@dlogin09:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
[gang.zhao@dlogin09:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -j 6
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1155.27 sec
2/6 Test #5: hafs_3denvar_hybens ..............   Passed  1586.49 sec
3/6 Test #4: hafs_4denvar_glbens ..............   Passed  1650.16 sec
4/6 Test #6: global_enkf ......................   Passed  1957.18 sec
5/6 Test #2: rtma .............................   Passed  3319.15 sec
6/6 Test #1: global_4denvar ...................   Passed  3546.50 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3546.52 sec
~~~
Here is the reports of the regression tests on Hera:
~~~
(work-env2) [Gang.Zhao@hfe01:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
(work-env2) [Gang.Zhao@hfe01:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -j 6
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 4: hafs_4denvar_glbens
    Start 1: global_4denvar
    Start 6: global_enkf
    Start 5: hafs_3denvar_hybens
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
1/6 Test #3: rrfs_3denvar_rdasens .............***Failed  2218.19 sec
2/6 Test #5: hafs_3denvar_hybens ..............   Passed  3687.92 sec
3/6 Test #2: rtma .............................   Passed  3828.48 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  4112.07 sec
5/6 Test #1: global_4denvar ...................   Passed  4118.91 sec
6/6 Test #6: global_enkf ......................   Passed  4119.37 sec

83% tests passed, 1 tests failed out of 6

Total Test time (real) = 4119.38 sec

The following tests FAILED:
          3 - rrfs_3denvar_rdasens (Failed)
Errors while running CTest
Output from these tests are in: /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
(work-env2) [Gang.Zhao@hfe01:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -R rrfs_3denvar_rdasens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 3: rrfs_3denvar_rdasens
1/1 Test #3: rrfs_3denvar_rdasens .............   Passed  619.88 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 619.90 sec
~~~

GangZhao-NOAA committed:
Adding code to analyze wave height and wind gust for 3DRTMA when using
ARW-ARW model forecast as firstguess.

GangZhao-NOAA committed:
if howv (or gust) is set in anavinfo, but its data is not found in
firstguess file, then the execution of code is aborted, instead of just
printing out the error message.

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/rtma3d_wavhgtwndgst_wrfarw

GangZhao-NOAA committed:
A trivial fix to a typo in comment.

TingLei-NOAA commented on src/gsi/rapidrefresh_cldsurf_mod.f90:
```diff
@@ -185,18 +187,25 @@ module rapidrefresh_cldsurf_mod
+!               contribution to the total BE of howv, so the total BE of howv is actually
+!               just the reduced static BE of howv. If to make the analysis of howv
+!               in hyrbid run is as similar as the analysis of howv in pure 3dvar run, 
+!               the static BE of howv used in hybrid run needs to be tuned (inflated actually).
```
@GangZhao-NOAA  Then, how can we do the inflation for static B for howv? Did I miss it?

GangZhao-NOAA replied:
@TingLei-NOAA 
Hi Ting, 

Sorry for my confusing comments in the source code.

What I mean is to give a larger background error for HOWV when running in hybrid EnVar analysis. 

For example, in the pure 3dvar run of GSI the background error of HOWV is 0.30 meters. If we would like to get the similar analysis of HOWV in a hybrid EnVar run of GSI with the weight for ensemble covariances = 50%, because there is no ensemble for HOWV yet, and if we want to get the analysis of HOWV as similar as from pure 3DVar, we could NOT use 0.3 meters as the static background error, because the static background error variances is reduced by the weight (100%-50%). To compensate for this reduction in background error, we need to "inflate" 0.3 to some larger values.
 
My explanation comment in the source code is really misleading, I will clean off these comments before merging the code.

Thank you very much for pointing this out!

-Gang

GangZhao-NOAA replied:
@TingLei-NOAA 
I removed these comment lines in the updated code.
Thank you!

TingLei-NOAA commented:
@GangZhao-NOAA  Thanks. Your inline comments are good and I just wondered if there are already existing options for user to do the inflation.  If there is no, that is ok. That comment will help users/developer understand the situation.

TingLei-NOAA reviewed (approved):
Thanks for continual enhancements of GSI

GangZhao-NOAA commented:
> @GangZhao-NOAA Thanks. Your inline comments are good and I just wondered if there are already existing options for user to do the inflation. If there is no, that is ok. That comment will help users/developer understand the situation.

@TingLei-NOAA Hi Ting, I realized that I should not use the word "inflat" here, since "covariance inflation" is somehow a specific term used in ensemble DA. I will change it to "increase" in my next PR. Since the process to increase the static error for HOWV and GUST are done outside of GSI code, not in GSI, so I'd better remove these comments in my next PR. Thank you again!

ShunLiu-NOAA commented on src/gsi/cplr_wrwrfmassa.f90:
```diff
@@ -2085,25 +2119,25 @@ subroutine wrwrfmassa_netcdf_wrf(this,mype)
        if ( laeroana_gocart ) then
           do iv = 1, n_gocart_var
-             i_chem(iv)=i_tt+(iv-1)*lm+1
+             i_chem(iv)=i_tt+(iv-1)*lm+1        ! ???? (tt is 3D field, should i_chem be i_tt + iv*lm ?)
```
do you still need the comment here?

GangZhao-NOAA replied:
@ShunLiu-NOAA 
Hi Shun,

This comment is not related to my modified code, and not necessary here. I am going to remove it  in the updated new code.

Thank you for pointing it out for me!

ShunLiu-NOAA commented on src/gsi/cplr_wrwrfmassa.f90:
```diff
@@ -2085,25 +2119,25 @@ subroutine wrwrfmassa_netcdf_wrf(this,mype)
        if ( wrf_pm2_5 ) then
           iv=1
-          i_chem(iv)=i_tt+(iv-1)*lm+1
+          i_chem(iv)=i_tt+(iv-1)*lm+1           ! ???? (tt is 3D field, should i_chem be i_tt + iv*lm ?)
```
same here

GangZhao-NOAA replied:
this comment will be removed.

RussTreadon-NOAA commented on src/gsi/cplr_read_wrf_mass_guess.f90:
```diff
@@ -1358,6 +1358,8 @@ subroutine read_wrf_mass_netcdf_guess_wrf(this,mype)
   !                            - add CV transform option on hydrometer variables
   !   2022-03-15  Hu  change all th2 to t2m and convert 2m temperature 
   !                        from potentionl to senseible temperature
+  !   2025-01-28  zhao   - added code to read wave height (howv) & wind gust (gust) 
+  !                        from intermediate binary file for the analysis in WRF-ARW based 3DRTMA
```
Adding changelog entries to source code is no longer needed.  Version control along with well documented issues record who changed the code and what the change is for.  

It's fine to keep the new changelog entries added by this PR. I suggest that future PRs do not add changelogs to the source code.

GangZhao-NOAA replied:
I removed this changelog entry as well as other new changelog entries introduced by this PR to GSI code. I will re-submit this PR soon.
Thank you for letting me know about this rule. I will keep it in mind.

RussTreadon-NOAA commented on src/gsi/cplr_regional_io.f90:
```diff
@@ -62,6 +62,8 @@ subroutine convert_regional_guess_wrf(this,mype,ctph0,stph0,tlm0)
   !   2005-05-24  pondeca - add 2dvar only surface analysis option
   !   2005-07-06  parrish - add variable update_pint
   !   2012-10-11  parrish - add byte_swap, which is set only on pe 0 and must be broadcast to all pes.
+  !   2025-01-29  zhao    - add i_howv_3dda/i_gust_3dda/i_howv_mask, which are set only on pe 0 and
```
Same comment as before.

GangZhao-NOAA replied:
new changelog entry is removed.

RussTreadon-NOAA commented on src/gsi/cplr_wrf_netcdf_interface.f90:
```diff
@@ -47,6 +47,8 @@ subroutine convert_netcdf_mass_wrf(this)
   !                            - add code for hydrometer variables needed for direct reflectivity DA
   !                            - add CV transform option on hydrometer variables for direct reflectivity DA
   !
+  !   2025-01-27  zhao   - added code for the input of wave height (howv) and wind gust (gust) fields from
```
Same comment as above.

GangZhao-NOAA replied:
this new changelog entry is removed.

RussTreadon-NOAA commented on src/gsi/cplr_wrf_netcdf_interface.f90:
```diff
@@ -1022,6 +1033,116 @@ subroutine convert_netcdf_mass_wrf(this)
+!            write(6,'(1x,A,A2,A,1x,I6)')trim(adjustl(myname_)),'::',trim(adjustl(derr_msg)),ierr
+!            call stop2(ierr)
+             call die(trim(adjustl(myname_)),trim(adjustl(derr_msg)),ierr)
+             i_howv_3dda = 0               ! skipping analysis of this variable
```
Question:  Doesn't `die` execute `mpi_abort` and terminate program execution?  If so,  execution will not reach the `i_howv_3dda = 0` line, right?   

Comment:  The commented out line `!      call stop2(ierr)` can be removed.

GangZhao-NOAA replied:
@RussTreadon-NOAA Hi Russ,
In mpeu_util.F90, "die" invokes subroutine "dropdead_", then in "dropdead_", it calls "mpi_abort" to terminate the GSI code (ifdef USE_MPI_ABORT, and it is defined), otherwise it calls "exit" to stop the running of GSI. 
Yes, here "die" terminates the execution. 

Yes, your right. Since the GSI code is going to be terminated under this condition, so it is not necessary to update "i_howv_3dda". The reason that I still keep this line here, is that if in the future it is suggested to keep the code running, instead of calling die, then the updated "i_howv_3dda" is ready to be broadcast to all PEs. So I will move the line "i_howv_3dda = 0" before call die in the updated new code.

Thank you!

RussTreadon-NOAA commented on src/gsi/cplr_wrf_netcdf_interface.f90:
```diff
@@ -1022,6 +1033,116 @@ subroutine convert_netcdf_mass_wrf(this)
+!            write(6,'(1x,A,A2,A,1x,I6)')trim(adjustl(myname_)),'::',trim(adjustl(derr_msg)),ierr
+!            call stop2(ierr)
+             call die(trim(adjustl(myname_)),trim(adjustl(derr_msg)),ierr)
+             i_gust_3dda = 0               ! skipping analysis of this variable
```
Similar to comment above.

GangZhao-NOAA replied:
line "i_gust_3dda = 0" is moved before call die.

RussTreadon-NOAA commented on src/gsi/cplr_wrf_netcdf_interface.f90:
```diff
@@ -2420,6 +2541,8 @@ subroutine update_netcdf_mass_wrf(this)
   !                            - add code for hydrometer variables needed for direct reflectivity DA
   !                            - add CV transform option on hydrometer variables for direct reflectivity DA
+  !   2025-01-27  zhao   - added code for the output of wave height (howv) and wind gust (gust) fields to
+  !                              the WRF-ARW (RAP/HRRR based) analysis file (netcdf fortmat).
```
Source code changelog is not necessary.

GangZhao-NOAA replied:
this changelog entry is removed.

RussTreadon-NOAA commented on src/gsi/unfill_mass_grid2.f90:
```diff
@@ -381,3 +381,107 @@ subroutine unfill_mass_grid2t_drycheck(gout,nx,ny,gin,qs)
+                            lakemask,i_howv_mask)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    unfill_mass_grid2t        opposite of fill_mass_grid2
```
This subroutine is `unfill_mass_grid2t_ldlkmask`, not `unfill_mass_grid2t`

GangZhao-NOAA replied:
@RussTreadon-NOAA 
Hi Russ,

I updated this information in documentation block. Thank you!

RussTreadon-NOAA commented on src/gsi/unfill_mass_grid2.f90:
```diff
@@ -381,3 +381,107 @@ subroutine unfill_mass_grid2t_drycheck(gout,nx,ny,gin,qs)
+!           but screen off analysis increment over land and/or lake . 
+!
+!
+! program history log:
```
If this is a new subroutine added to GSI, why does it have a program history going back to 2004?

GangZhao-NOAA replied:
@RussTreadon-NOAA,
I updated this information. Thank you!

RussTreadon-NOAA commented on src/gsi/unfill_mass_grid2.f90:
```diff
@@ -381,3 +381,107 @@ subroutine unfill_mass_grid2t_drycheck(gout,nx,ny,gin,qs)
+! Only add analysis increment over sea water (no land and lake)
+  select case (i_howv_mask)
+    case(2)
+      write(6,*) 'unfill_mass_grid2t_ldlkmask: using landlakemask to screen off HOWV analysis increment over land and lake.'
```
Do multiple mpi tasks call this routine?  If so, `write(6,*)` could generate a lot of output.  Do we need this output from all tasks?  Should we place write statements in this routine inside verbose blocks?

GangZhao-NOAA replied:
@RussTreadon-NOAA 
Hi Russ,

I double checked the code, in GSI this subroutine is only invoked on **pe 0** . So it should not be called by multiple MPI tasks simultaneously.
in subroutine **wrwrfmassa_netcdf_wrf**(this,mype)
`...`
`      if(mype == 0) then` 
`...`
`          if ( i_howv_mask > 0 ) then`
`             call unfill_mass_grid2t_ldlkmask(tempa,im,jm,temp1,landmask, &`                                         
`                                                            lakemask,i_howv_mask)`
`...`
`   end if     !endif mype==0`
`...`

Thank you!

RussTreadon-NOAA reviewed (commented):
Did not test code.  Most comments are minor.   

Question:  Do any of the existing GSI ctests exercise the functionality added by this PR?  If not, (a) how do we know the changes work as intended? and (b) what's the strategy to ensure future PRs do not break the functionality added by this PR?

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/rtma3d_wavhgtwndgst_wrfarw

GangZhao-NOAA commented on src/gsi/cplr_wrwrfmassa.f90:
```diff
@@ -1850,6 +1850,8 @@ subroutine wrwrfmassa_netcdf_wrf(this,mype)
   !                        from sensible to potentionl temperature for writing
   !   2022-03-15  Hu  add a check to remove the negative moisture if rh>4%
   !                   for the lowest 3 model level
+  !   2025-01-28  zhao   - added code to write out analyses of wave height (howv) & wind gust (gust)
```
Following the suggestion from Russ, this new changelog entry is not necessary and will be removed in the updated code.

GangZhao-NOAA commented on src/gsi/rapidrefresh_cldsurf_mod.f90:
```diff
@@ -33,6 +33,8 @@ module rapidrefresh_cldsurf_mod
 !                         corp_howv:   to set the static background error of howv
 !                         hwllp_howv:  to set the de-correlation length scale
 !                         i_howv_3dda: control the analysis of howv in 3D analysis (if howv is in anavinfo)
+!   2025-01-28 zhao  -- added option i_howv_mask to control the screen off the wave height
```
this new changelog entry will be removed.

GangZhao-NOAA commented on src/gsi/setuphowv.f90:
```diff
@@ -33,6 +33,8 @@ subroutine setuphowv(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diag
 !                       . Remove my_node with corrected typecast().
 !   2018-01-08  pondeca - addd option l_closeobs to use closest obs to analysis
 !                                     time in analysis
+!   2025-01-24  zhao    - add option l_obsprvdiag to control the dump-out of obs provider
```
remove this new changelog entry.

GangZhao-NOAA commented:
> Did not test code. Most comments are minor.
> 
> Question: Do any of the existing GSI ctests exercise the functionality added by this PR? If not, (a) how do we know the changes work as intended? and (b) what's the strategy to ensure future PRs do not break the functionality added by this PR?

@RussTreadon-NOAA 
Hi Russ, 

> _Did not test code. Most comments are minor._

This code was tested with a real case data (2024-12-12_00z) 

> _Question: Do any of the existing GSI ctests exercise the functionality added by this PR?_ 

As you pointed out, NONE of the existing GSI ctests exercises the functionality added by this PR. 

> _If not, (a) how do we know the changes work as intended?_

This code was tested with a real case data (2024-12-12_00z) with the added functionality (analyzing wave height and wind gust in hybrid 3DEnVar with HRRR-based first guess). And it passed all the 6 existing ctests (regression test) in current GSI package.

> _(b) what's the strategy to ensure future PRs do not break the functionality added by this PR?_

Thank you for pointing out this question!  I discussed with Manuel (@ManuelPondeca-NOAA). We should add a new ctest item for the new functionalities added for 3DRTMA. We are busy with current HRRR-3DRTMA workflow project, so we cannot provide such kind of ctest for now. We will do it in the near future.

GangZhao-NOAA committed:
Clear off the new changelog entries introduced by PR#835, and a few
minor changes suggested by code reviewers.
With version control on github along with well documented
issues, added changelog entries to source code is no longer needed.

GangZhao-NOAA commented:
@guoqing-noaa @TingLei-NOAA @ShunLiu-NOAA @RussTreadon-NOAA
Hi Guoqing, Ting, Shun and Russ,

Following your questions/suggestions/comments, I modified the code and re-submitted the PR #835 for review. Most of the updated part are in the comments, so there should be no essential differences, comparing to the previous commit. If you have time, could you please make another round code review on these newly-updated code?

I am running all the 6 existing ctests of GSI on Hera now, and will run ctests on WCOSS2 very soon this afternoon. Then I will update the new reports from the ctests here.

Thank you very much for your help!

-Gang

RussTreadon-NOAA reviewed (commented):
Thank you @GangZhao-NOAA for updating the code.  Changes look good.

GangZhao-NOAA commented:
While this PR is under review, the GSI main branch (develop) was updated to [commit #8e54642](https://github.com/NOAA-EMC/GSI/commit/8e546427716d5913c746bcfaf9d89e6d82b90b49). So I merged that latest update into my code, and then following the suggestions & comments from reviewers, made some modifications to the code. Now this PR #835 is to merge my  latest [commit #94bed49](https://github.com/GangZhao-NOAA/GSI/commit/94bed49b14d7047db4dac961f84e0637b56d6318) (in branch "feature/rtma3d_wavhgtwndgst_wrfarw") to EMC-GSI.  Due to these new changes to the code, I re-ran the ctests on Hera and WCOSS2 (Cactus):
- [x] Regression tests on WCOSS2 (Cactus) and Hera: my latest GSI commit [#94bed49](https://github.com/GangZhao-NOAA/GSI/commit/94bed49b14d7047db4dac961f84e0637b56d6318)  vs last EMC-GSI commit [#8e54642](https://github.com/NOAA-EMC/GSI/commit/8e546427716d5913c746bcfaf9d89e6d82b90b49)
Here is the reports of the ctest (regression tests) on Hera:
~~~
(work-env2) [Gang.Zhao@hfe04:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
(work-env2) [Gang.Zhao@hfe04:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -j 6
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  681.04 sec
2/6 Test #2: rtma .............................   Passed  2233.63 sec
3/6 Test #6: global_enkf ......................   Passed  2467.64 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  2486.48 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  2607.95 sec
6/6 Test #1: global_4denvar ...................   Passed  3900.66 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3900.68 sec
~~~

Here is the reports of the ctest (regression tests) running on WCOSS2 (Cactus):
~~~
[gang.zhao@clogin01:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
[gang.zhao@clogin01:build] (feature/rtma3d_wavhgtwndgst_wrfarw)$ ctest -j 6
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  919.52 sec
2/6 Test #5: hafs_3denvar_hybens ..............   Passed  1356.76 sec
3/6 Test #2: rtma .............................   Passed  1641.96 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  1648.40 sec
5/6 Test #6: global_enkf ......................   Passed  1678.64 sec
6/6 Test #1: global_4denvar ...................   Passed  2286.87 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 2286.96 sec
~~~

RussTreadon-NOAA reviewed (approved):
Ctests pass on WCOSS2 and Hera.

Approve.

ShunLiu-NOAA commented:
> Ctests pass on WCOSS2 and Hera.
> 
> Approve.

Russ, Thank you for running ctest on WCOSS2 and Hera.

ShunLiu-NOAA commented:
@TingLei-NOAA Could you review the new changes of this PR?

EdwardColon-NOAA reviewed (approved):
The issues related to code documentation and potentially excessive print calls have previously been resolved. Approve.

RussTreadon-NOAA commented:
> > Ctests pass on WCOSS2 and Hera.
> > Approve.
> 
> Russ, Thank you for running ctest on WCOSS2 and Hera.

Credit goes to @GangZhao-NOAA for rerunning ctests on WCOSS2 and Hera.  He also ran the first round of ctests.  Thank you @GangZhao-NOAA for running the tests again.  It was good to confirm that the code still behaves as expected after you integrated reviewers' input.

RussTreadon-NOAA commented:
For the sake of completeness run ctests on Orion and Hercules.  Results are posted below

Orion
```
Test project /work2/noaa/da/rtreadon/git/gsi/pr835/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #6: global_enkf ......................   Passed  1027.76 sec
2/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1147.22 sec
3/6 Test #2: rtma .............................   Passed  1928.00 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  3022.20 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  3200.51 sec
6/6 Test #1: global_4denvar ...................   Passed  3721.97 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3721.98 sec
```

Hercules
```
Test project /work/noaa/da/rtreadon/git/gsi/pr835/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  1865.70 sec
2/6 Test #6: global_enkf ......................   Passed  1925.50 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  2292.33 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  2842.59 sec
5/6 Test #2: rtma .............................   Passed  2885.18 sec
6/6 Test #1: global_4denvar ...................   Passed  3061.19 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 3061.20 sec
```

All ctests pass on both MSU machines.

GangZhao-NOAA commented:
@RussTreadon-NOAA 
Hi Russ,
Thank you so much for running the ctests on MSU machines and making the tests completeness! I am so grateful!
-Gang
